src/sentry/analytics/event.py

```diff
@@ -14,6 +16,32 @@ def __init__(self, name, type=six.text_type, required=True):
         self.required = required
 
 
+class Map(Attribute):
+    def __init__(self, name, attributes, required=True):
+        self.name = name
+        self.required = required
+        self.attributes = attributes
+
+    def type(self, value):
+        if not isinstance(value, dict):
+            raise ValueError('Value must be a dictionary')
+
+        data = {}
+        for attr in self.attributes:
+            if attr.required:
+                nv = value.pop(attr.name)
```

Comment:
Mutating the input value here seems to violate the principle of least surprise and will almost certainly lead to bugs if the input value is not used exclusively for this method